examples/reporting/parallel_report_download.py

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Shows how to download in parallel a set of reports from a list of accounts.
+
+If you need to obtain a list of accounts, please see the
+account_management/get_account_hierarchy.py or
+account_management/list_accessible_customers.py examples.
+"""
+
+import argparse
+from itertools import product
+import multiprocessing
+import sys
+import time
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+# Maximum number of processes to spawn.
+MAX_PROCESSES = multiprocessing.cpu_count()
+# Timeout between retries in seconds.
+BACKOFF_FACTOR = 5
+# Maximum number of retries for errors.
+MAX_RETRIES = 5
+
+
+def main(client, customer_ids):
+    """The main method that creates all necessary entities for the example.
+
+    Args:
+        client: an initialized GoogleAdsClient instance.
+        customer_ids: an array of client customer IDs.
+    """
+
+    # Define the GAQL query strings to run for each customer ID.
+    campaign_query = """
+        SELECT campaign.id, metrics.impressions, metrics.clicks
+        FROM campaign
+        WHERE segments.date DURING LAST_30_DAYS"""
+    ad_group_query = """
+        SELECT campaign.id, ad_group.id, metrics.impressions, metrics.clicks
+        FROM ad_group
+        WHERE segments.date DURING LAST_30_DAYS"""
+
+    inputs = _generate_inputs(
+        client, customer_ids, [campaign_query, ad_group_query]
+    )
+    with multiprocessing.Pool(MAX_PROCESSES) as pool:
+        # Call _issue_search_request on each input, parallelizing the work
+        # across processes in the pool.
+        results = pool.starmap(_issue_search_request, inputs)
+
+        # Partition our results into successful and failed results.
+        successes = []
+        failures = []
+        for res in results:
```

Comment:
The issue here is that there is no context for the success or failure. In the case of a failure, how do you know which input failed?

Reply:
That would be easy to add to the object that gets returned to _issue_search_request. In addition to "exception", there could be a field "data".

Reply:
Updated -- lmk if that's what you had in mind.